eth/downloader/queue.go

```diff
@@ -319,49 +483,86 @@ func (q *queue) Reserve61(p *peer, count int) *fetchRequest {
+func (q *queue) ReserveReceipts(p *peer, count int) (*fetchRequest, bool, error) {
+	noop := func(header *types.Header) bool {
+		return header.ReceiptHash == types.EmptyRootHash
+	}
+	return q.reserveHeaders(p, count, q.receiptTaskPool, q.receiptTaskQueue, q.receiptPendPool, q.receiptDonePool, noop)
+}
+
+// reserveHeaders reserves a set of data download operations for a given peer,
+// skipping any previously failed ones. This method is a generic version used
+// by the individual special reservation functions.
+func (q *queue) reserveHeaders(p *peer, count int, taskPool map[common.Hash]*types.Header, taskQueue *prque.Prque,
+	pendPool map[string]*fetchRequest, donePool map[common.Hash]struct{}, noop func(*types.Header) bool) (*fetchRequest, bool, error) {
 	q.lock.Lock()
 	defer q.lock.Unlock()
 
 	// Short circuit if the pool has been depleted, or if the peer's already
 	// downloading something (sanity check not to corrupt state)
-	if q.headerQueue.Empty() {
+	if taskQueue.Empty() {
 		return nil, false, nil
 	}
-	if _, ok := q.pendPool[p.id]; ok {
+	if _, ok := pendPool[p.id]; ok {
 		return nil, false, nil
 	}
-	// Calculate an upper limit on the bodies we might fetch (i.e. throttling)
-	space := len(q.blockCache) - len(q.blockPool)
-	for _, request := range q.pendPool {
+	// Calculate an upper limit on the items we might fetch (i.e. throttling)
+	space := len(q.resultCache) - len(donePool)
+	for _, request := range pendPool {
 		space -= len(request.Headers)
 	}
-	// Retrieve a batch of headers, skipping previously failed ones
+	// Retrieve a batch of tasks, skipping previously failed ones
 	send := make([]*types.Header, 0, count)
 	skip := make([]*types.Header, 0)
 
-	process := false
-	for proc := 0; proc < space && len(send) < count && !q.headerQueue.Empty(); proc++ {
-		header := q.headerQueue.PopItem().(*types.Header)
+	progress := false
+	for proc := 0; proc < space && len(send) < count && !taskQueue.Empty(); proc++ {
+		header := taskQueue.PopItem().(*types.Header)
 
-		// If the header defines an empty block, deliver straight
-		if header.TxHash == types.DeriveSha(types.Transactions{}) && header.UncleHash == types.CalcUncleHash([]*types.Header{}) {
-			if err := q.enqueue("", types.NewBlockWithHeader(header)); err != nil {
-				return nil, false, errInvalidChain
+		// If we're the first to request this task, initialize the result container
+		index := int(header.Number.Int64() - int64(q.resultOffset))
+		if index >= len(q.resultCache) || index < 0 {
+			return nil, false, errInvalidChain
+		}
+		if q.resultCache[index] == nil {
+			components := 1
+			if q.mode == FastSync && header.Number.Uint64() <= q.fastSyncPivot {
+				components = 2
 			}
-			delete(q.headerPool, header.Hash())
-			process, space, proc = true, space-1, proc-1
+			q.resultCache[index] = &fetchResult{
+				Pending: components,
+				Header:  header,
+			}
+		}
+		// If this fetch task is a noop, skip this fetch operation
+		if noop(header) {
```

Comment:
`skipFetch` doesn't really convey the reason it should be skipped. Noop imho better signals that you don't *need* to fetch because *no data* will be returned (hence the noop name).